### ISSUE TITLE
sql: panic in EncodeSQLStringWithFlags

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#30474]
https://sentry.io/cockroach-labs/cockroachdb/issues/695696446/

```
*log.safeError: conn_executor.go:668: panic while executing 1 statements: UPDATE _ SET _ = $1, _ = $2, _ = _ WHERE (_ = $3) AND (_ = $4): caused by <redacted>
  File "github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go", line 414, in func1
  File "github.com/cockroachdb/cockroach/pkg/sql/lex/encode.go", line 112, in EncodeSQLStringWithFlags
  File "github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go", line 965, in Format
  File "github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go", line 47, in formatNodeOrHideConstants
  File "github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go", line 276, in FormatNode
...
(18 additional frame(s) were not displayed)

conn_executor.go:668: panic while executing 1 statements: UPDATE _ SET _ = $1, _ = $2, _ = _ WHERE (_ = $3) AND (_ = $4): caused by <redacted>
```

**Comment by knz** [comment]
Full stack trace:

```
    github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go in func1 at line 414
    Called from: runtime/asm_amd64.s in call32
    github.com/cockroachdb/cockroach/pkg/sql/lex/encode.go in EncodeSQLStringWithFlags at line 112
    github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go in Format at line 965
    github.com/cockroachdb/cockroach/pkg/sql/sem/tree/hide_constants.go in formatNodeOrHideConstants at line 47
    github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go in FormatNode at line 276
    github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go in AsStringWithFlags at line 306
    github.com/cockroachdb/cockroach/pkg/sql/targets.go in symbolicExprStr at line 167
    github.com/cockroachdb/cockroach/pkg/sql/targets.go in addOrReuseRenderStartingFromIdx at line 122
    github.com/cockroachdb/cockroach/pkg/sql/targets.go in addOrReuseRender at line 112
    github.com/cockroachdb/cockroach/pkg/sql/update.go in addOrMergeExpr at line 752
    github.com/cockroachdb/cockroach/pkg/sql/update.go in Update at line 310
    github.com/cockroachdb/cockroach/pkg/sql/plan.go in doPrepare at line 943
    github.com/cockroachdb/cockroach/pkg/sql/plan.go in prepare at line 851
    github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go in func1 at line 218
    github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go in prepare at line 238
    github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go in addPreparedStmt at line 122
    github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go in execPrepare at line 55
    github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go in run at line 1143
    github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go in ServeConn at line 416
    github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go in func3 at line 267
```

**Comment by knz** [comment]
Error here:

```go
func EncodeSQLStringWithFlags(buf *bytes.Buffer, in string, flags EncodeFlags) {
  // See http://www.postgresql.org/docs/9.4/static/sql-syntax-lexical.html
  start := 0
  escapedString := false
  bareStrings := flags.HasFlags(EncBareStrings)
  // Loop through each unicode code point.
  for i, r := range in {
    ch := byte(r)
    if r >= 0x20 && r < 0x7F {
      if mustQuoteMap[ch] {
        // We have to quote this string - ignore bareStrings setting
        bareStrings = false
      }
      if !stringencoding.NeedEscape(ch) && ch != '\'' {
        continue
      }
    }

    if !escapedString {
      buf.WriteString("e'") // begin e'xxx' string
      escapedString = true
    }
    buf.WriteString(in[start:i])  // <---- HERE
    ln := utf8.RuneLen(r)
    if ln < 0 {
      start = i + 1
    } else {
      start = i + ln
    }
    stringencoding.EncodeEscapedChar(buf, in, r, ch, i, '\'')
  }
```

**Comment by knz** [comment]
Ok by looking at the implementation of `utf8.RuneLen` and `runtime.decoderune` (used by `range`) side-by-side I found they do not match, so this code is wrong on its face. It was easy to find something suspicious easily:

```
> select quote_literal(b'abc\xd8\xffdef'::string);
     quote_literal
+----------------------+
  e'abc\xd8\xff\x64ef'
(1 row)
```

The 3 first and 3 last characters are ASCII and thus ought to be printed as-is, however the encode logic gets confused and prints "d" as `\x64` instead. This is because RuneLen on the surrogate char (\xd8\xff) returns -1, whereas `runtime.decoderune` properly skips 2 bytes ahead.

There may be other inconsistencies in there but I would like to suggest that any Go code that mixes string range iterations with `utf8.RuneLen` be declared outright suspicious. 

Luckily we only have 2 such uses right now, one here and one in `stringencoding.EncodeEscapedChar` so the problem is well contained.

**Comment by knz** [comment]
It *really* does not help that `utf8.DecodeRune` and `runtime.decoderune` don't do the same thing...

**Comment by robert-s-lee** [comment]
This is the actual SQL that caused the issue from linkbench test.

```
* ERROR: [n1,client=127.0.0.1:51903,user=root] a SQL panic has occurred while executing "UPDATE nodetable SET version = $1, \"time\" = $2, data = b'r~\\x89\\x8e\\xb7\\x98p\\xd6\   \xa2\\x9fl\\xabtEae;\\xbf\\xc3@p9\\xb7\\xa1P\\xc2\\xa0Z\\xdc\\xba\\x82S\\xa7\\x8c\\xb5;8\\xcf\\xd9R\\x8e\\xa4\\xcds\\xadt\\xa3AM7\\xaaAGv@S`~dxCHN\\x8c\\xc6Cm\\x89D\\xb176\\xdc6   \\x9c\\x88\\xcf\\xd9\\xa0B\\xa59H\\x92c\\xd1\\xd9bR\\xaa\\x9e\\\\\\x93\\xbaq\\x93\\xa1p\\x9c\\xa5\\xb9\\x90\\xdc\\x8f\\xd3\\x93\\xa5\\xb6B\\xa3\\xc8f`3\\x8a\\xa1v\\xcfA\\xd2;^\\   x917\\x9d\\xd5|nc\\xccy\\xd7mT\\xda\\x81N\\x84V\\x8a?\\xc3^?@?\\xc9~H\\xa4r\\x88F\\xd9\\xa3k\\xb2d\\x84Z' WHERE (id = $3) AND (type = $4)": runtime error: slice bounds out of range
*
*
* ERROR: [n1,client=127.0.0.1:51903,user=root] a panic has occurred!
*
panic while executing 1 statements: UPDATE _ SET _ = $1, _ = $2, _ = _ WHERE (_ = $3) AND (_ = $4); caused by runtime error: slice bounds out of range

* ERROR: [n1,client=127.0.0.1:51903,user=root] Reported as error ef62dd44f1c048a2892c54ebd99288b0
```

**Comment by knz** [comment]
@robert-s-lee there is not enough information to re-construct the original SQL query and, more importantly, the data string. Any chance you know where this data came from?

**Comment by knz** [comment]
(I don't mean the value of the `data` column in the table. That one is fine. I'd like to see the data strings for all the values)

**Comment by robert-s-lee** [comment]
For re-pro,  I am using OSX, java version "1.8.0_131", 1 node CRDB.  

The data load will produce panic immediately.  The 2nd part of the workload I am still working to port.

```
git clone https://github.com/robert-s-lee/oltpbench.git --branch cockroachdb --single-branch oltpbenchmark 
cd oltpbenchmark
ant
./oltpbench -w linkbench -l
```

**Comment by jordanlewis** [comment]
@arjunravinarayan didn't you poke at something similar recently? maybe you have an idea of what's going on here.

**Comment by arjunravinarayan** [comment]
@knz I fixed a panic in #29549 that was backported in #29668. That didn't make it into the reported beta (20180904). It could be the same issue.

**Comment by knz** [comment]
yep that looks like it!

**Comment by arjunravinarayan** [comment]
OK, let's call it closed by #29549 then. Without a more concrete query, it's hard to reproduce and be 100% sure, but this is the best understanding we have at this time.
